### PR TITLE
Finance dashboard

### PR DESCRIPTION
**Pull request by jb3**
Adds a new finance dashboard requested by Alfie which should allow for some interesting figures to be presented in the upcoming General Meeting.

image

This follows the same RBAC as other routes within the invoices section of the application so only trusted users can view this information.

Time periods allow for figures to be viewed for all time, for the last week, month or year.

The above screenshot misattributes payment sources as they are not part of the seeded data generated by the bootstrap script, though this data is accurately attributed in production.

### CODE REVIEW

**Comment by FreneticScribbler** [comment]
Oh that's *$%ing sexy. Damn. 

The layout does feel a bit 'tablet' to me, but that's a personal preference for sure.  

I'll give it the in depth look it deserves shortly, but :star:. Nice.

**Comment by FreneticScribbler** [comment]
Feels like we need better test data finance wise now! :laughing:

File: RIGS/views/finance.py [hunk 1]
```diff
@@ -18,8 +18,76 @@
...
+        for method in context['payment_methods']:
+            method['method'] = payment_methods.get(method['method'], f"Unknown method ({method['method']})")
+
+        context["total_income"] = sum([i['total'] for i in context['payment_methods']])
+
+        payments = context['payments']
+        mean_duration = 0
+
+        for payment in payments:
+            mean_duration += days_between(payment.invoice.invoice_date, payment.date)
```
Comment (FreneticScribbler) [inline RIGS/views/finance.py:80]: Can you not just subtract invoice_date from payment_date and then do .days on the resuling timedelta? Why does it need to be converted to seconds? Think I'm missing something here

Comment (jb3) [reply RIGS/views/finance.py:80]: This method allows for fractional days to be counted, I agree we could do that and get the same result but this method is more accurate and for testing purposes works a bit nicer.

The `days` property is always an integer, not a float.

Comment (FreneticScribbler) [reply RIGS/views/finance.py:80]: Ah I'm with you now :+1:

**Review by FreneticScribbler — APPROVED** [review]

**Comment by FreneticScribbler** [comment]
Now watch the stochastic test suite fail on the second run through...

**Comment by FreneticScribbler** [comment]
Was seeing a critical application error loading the dashboard yesterday - went to investigate today and it appears to no longer be occurring? Weird...

**Comment by jerbzz** [comment]
Stochastic in all respects

**Comment by jb3** [comment]
> Was seeing a critical application error loading the dashboard yesterday - went to investigate today and it appears to no longer be occurring? Weird... 

I think this could be because I made the great application design decision of defaulting to _all_ payments ever.

I'm going to swap it around to default to the smallest time period first and then I'll look at some query optimisation, this was just a starting point and there are some optimisations to be made to the ORM-generated SQL.

**Comment by FreneticScribbler** [comment]
Average time to pay seems to be showing a _slightly_ questionable number on the live data, as well?

**Comment by jb3** [comment]
> Average time to pay seems to be showing a _slightly_ questionable number on the live data, as well?

This was also noticed the other day with Alfie. I think this is a technicality that internal SU payments are all attributed to the same day.

I'm going to break it down by payment method to hopefully get some more sensible data.

**Pull request merged**